src/cache.ts

```diff
@@ -0,0 +1,229 @@
+		return cachedValue.value;
+	}
+}
+
+export const DEFAULT_R2_BUCKET_CACHE_TTL_IN_MS = 5 * 60 * 1000;
+
+export class CachedR2Object {
+	public checksums: R2Checksums;
+	public customMetadata?: Record<string, string>;
+	public etag: string;
+	public httpEtag: string;
+	public httpMetadata?: R2HTTPMetadata;
+	public key: string;
+	public size: number;
+	public uploaded: Date;
+	public version: string;
+
+	constructor(
+		object: R2Object,
+		public data?: Uint8Array
+	) {
+		this.checksums = object.checksums;
+		this.customMetadata = object.customMetadata;
+		this.etag = object.etag;
+		this.httpEtag = object.httpEtag;
+		this.httpMetadata = object.httpMetadata;
+		this.key = object.key;
+		this.size = object.size;
+		this.uploaded = object.uploaded;
+		this.version = object.version;
+	}
+}
+
+export class CachedR2Objects {
+	public delimitedPrefixes: string[];
+	public objects: CachedR2Object[];
+	public truncated: boolean;
+
+	constructor(objects: R2Objects) {
+		this.delimitedPrefixes = objects.delimitedPrefixes;
+		this.objects = objects.objects.map(o => new CachedR2Object(o));
+		this.truncated = objects.truncated;
+	}
+}
+
+const R2Method = {
+	GET: 'get',
+	HEAD: 'head',
+	LIST: 'list',
+};
+
+export class CachedR2Bucket {
+	constructor(
+		private ctx: Context,
+		private bucket: R2Bucket,
+		private cache: ReadableCache,
+		private ttl_in_ms = DEFAULT_R2_BUCKET_CACHE_TTL_IN_MS
+	) {}
+
+	// WARNING: key should be lowered than 1024 bytes
+	// See https://developers.cloudflare.com/r2/reference/limits/
+	head(key: string): Promise<CachedR2Object | null> {
+		const cacheKey = `head/${key}`;
+		return this.cache.read(cacheKey, async () => {
+			this.ctx.metrics.r2RequestsTotal.inc({ method: R2Method.HEAD });
+			const object = await this.bucket.head(key);
+			if (object === null) {
+				return { value: null, expiration: new Date() };
+			}
+			const value = new CachedR2Object(object);
+			return {
+				value,
+				expiration: new Date(Date.now() + this.ttl_in_ms),
+			};
+		});
+	}
+
+	list(options?: R2ListOptions | undefined): Promise<CachedR2Objects> {
+		const cacheKey = `list/${JSON.stringify(options)}`;
+		return this.cache.read(cacheKey, async () => {
+			this.ctx.metrics.r2RequestsTotal.inc({ method: R2Method.LIST });
+			const objects = await this.bucket.list(options);
+			const value = new CachedR2Objects(objects);
+			return {
+				value,
+				expiration: new Date(Date.now() + this.ttl_in_ms),
+			};
+		});
+	}
+
+	// WARNING: key should be lowered than 1024 bytes
+	// See https://developers.cloudflare.com/r2/reference/limits/
+	async get(key: string, options?: R2GetOptions): Promise<CachedR2Object | null> {
+		const cacheKey = `get/${key}`;
+		return this.cache.read(cacheKey, async () => {
+			this.ctx.metrics.r2RequestsTotal.inc({ method: R2Method.GET });
+			const object = await this.bucket.get(key, options);
+			if (object === null) {
+				return { value: null, expiration: new Date() };
+			}
```

Comment:
shouldn't this return `null`?

Reply:
I see it does satisfy the interface, I have no recommendation for something different.